realm/realm-library/src/syncIntegrationTest/kotlin/io/realm/SyncedRealmIntegrationTests.kt

```diff
@@ -0,0 +1,308 @@
+import io.realm.exceptions.RealmMigrationNeededException
+import io.realm.internal.OsRealmConfig
+import io.realm.kotlin.syncSession
+import io.realm.log.LogLevel
+import io.realm.log.RealmLog
+import io.realm.log.RealmLogger
+import io.realm.mongodb.*
+import io.realm.mongodb.sync.*
+import io.realm.rule.BlockingLooperThread
+import io.realm.util.assertFailsWithErrorCode
+import org.bson.BsonObjectId
+import org.bson.types.ObjectId
+import org.junit.*
+import org.junit.Assert.*
+import org.junit.runner.RunWith
+import java.io.File
+import java.util.*
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+private const val SECRET_PASSWORD = "123456"
+
+/**
+ * Catch all class for tests that not naturally fit anywhere else.
+ */
+@RunWith(AndroidJUnit4::class)
+class SyncedRealmIntegrationTests {
+
+    private val looperThread = BlockingLooperThread()
+
+    private lateinit var app: App
+    private lateinit var user: User
+    private lateinit var syncConfiguration: SyncConfiguration
+
+    private val configurationFactory: TestSyncConfigurationFactory = TestSyncConfigurationFactory()
+
+    @Before
+    fun setup() {
+        Realm.init(InstrumentationRegistry.getInstrumentation().targetContext)
+        RealmLog.setLevel(LogLevel.ALL)
+        app = TestApp()
+        user = app.registerUserAndLogin(TestHelper.getRandomEmail(), SECRET_PASSWORD)
+        syncConfiguration = configurationFactory
+                // TODO We generate new partition value for each test to avoid overlaps in data. We
+                //  could make test booting with a cleaner state by somehow flushing data between
+                //  tests.
+                .createSyncConfigurationBuilder(user, BsonObjectId(ObjectId()))
+                .modules(DefaultSyncSchema())
+                .build()
+    }
+
+    @After
+    fun teardown() {
+        if (this::app.isInitialized) {
+            app.close()
+        }
+        RealmLog.setLevel(LogLevel.WARN)
+    }
+
+    @Test
+    fun loginLogoutResumeSyncing() = looperThread.runBlocking {
+        val config: SyncConfiguration = configurationFactory.createSyncConfigurationBuilder(user, user.id)
+                .testSchema(SyncStringOnly::class.java)
+                .testSessionStopPolicy(OsRealmConfig.SyncSessionStopPolicy.IMMEDIATELY)
+                .build()
+        Realm.getInstance(config).use { realm ->
+            realm.executeTransaction {
+                realm.createObject(SyncStringOnly::class.java, ObjectId()).chars = "Foo"
+            }
+            realm.syncSession.uploadAllLocalChanges()
+            user.logOut()
+        }
+        try {
+            assertTrue(Realm.deleteRealm(config))
+        } catch (e: IllegalStateException) {
+            // FIXME: We don't have a way to ensure that the Realm instance on client thread has been
```

Comment:
Hmm, I think I saw something about this being fixed. Try talking to @RedBeard0531

Reply:
The old issue #5416 is still open. I have tried incorporating the `shutdown_and_wait` from https://github.com/realm/realm-object-store/pull/1055 but had various issue around it. In some of the other tests referring to #5416 I was not able to get hold of the session and other places like here it did not work out. Reapplied a workaround from another test by retried to close if it fails.

Reply:
Maybe related to some internal notification triggering to early due to https://github.com/realm/realm-object-store/pull/1054